KOGITO-7765: Example with Serverless workflow with timeouts and job-service

wmedvede opened the pull request:
- Initial version

Many thanks for submitting your Pull Request :heart:! 

Please make sure that your PR meets the following requirements:

**WARNING! Please make sure you are opening your PR against `main` branch!**

- [X] You have read the [contributors guide](https://github.com/kiegroup/kogito-runtimes#contributing-to-kogito)
- [X] Pull Request title is properly formatted: `KOGITO-XYZ Subject`
- [X] Pull Request title contains the target branch if not targeting main: `[0.9.x] KOGITO-XYZ Subject`
- [X] Pull Request contains link to the JIRA issue
- [X] Pull Request contains link to any dependent or related Pull Request
- [X] Pull Request contains description of the issue
- [X] Pull Request does not include fixes for issues other than the main ticket

<details>
<summary>
How to replicate CI configuration locally?
</summary>

Build Chain tool does "simple" maven build(s), the builds are just Maven commands, but because the repositories relates and depends on each other and any change in API or class method could affect several of those repositories there is a need to use [build-chain tool](https://github.com/kiegroup/github-action-build-chain) to handle cross repository builds and be sure that we always use latest version of the code for each repository.
 
[build-chain tool](https://github.com/kiegroup/github-action-build-chain) is a build tool which can be used on command line locally or in Github Actions workflow(s), in case you need to change multiple repositories and send multiple dependent pull requests related with a change you can easily reproduce the same build by executing it on Github hosted environment or locally in your development environment. See [local execution](https://github.com/kiegroup/github-action-build-chain#local-execution) details to get more information about it.
</details>

<details>
<summary>
How to retest this PR or trigger a specific build:
</summary>

- for <b>pull request checks</b>  
  Please add comment: <b>Jenkins retest this</b>

- for a <b>specific pull request check</b>  
  Please add comment: <b>Jenkins (re)run [kogito-examples] tests</b>

- for <b>quarkus branch checks</b>  
  Run checks against Quarkus current used branch  
  Please add comment: <b>Jenkins run quarkus-branch</b>

- for a <b>quarkus branch specific check</b>  
  Run checks against Quarkus current used branch  
  Please add comment: <b>Jenkins (re)run [kogito-examples] quarkus-branch</b>

- for <b>quarkus main checks</b>  
  Run checks against Quarkus main branch  
  Please add comment: <b>Jenkins run quarkus-main</b>

- for a <b>specific quarkus main check</b>  
  Run checks against Quarkus main branch  
  Please add comment: <b>Jenkins (re)run [kogito-examples] quarkus-branch</b>

- for <b>native checks</b>  
  Run native checks  
  Please add comment: <b>Jenkins run native</b>

- for a <b>specific native check</b>  
  Run native checks 
  Please add comment: <b>Jenkins (re)run [kogito-examples] native</b>

- for <b>mandrel checks</b>  
  Run native checks against Mandrel image
  Please add comment: <b>Jenkins run mandrel</b>

- for a <b>specific mandrel check</b>  
  Run native checks against Mandrel image  
  Please add comment: <b>Jenkins (re)run [kogito-examples] mandrel</b>

</details>

wmedvede commented:
Hi @ricardozanini , @tiagodolphine ,
here goes the initial version of the Serverless Workflows Timeouts Showcase. Although I'll continue doing improvements, the core installation procedure and the workflows are in place.

It'll be great if you can give it a try locally to check the procedure, etc.

Thanks!

ricardozanini commented on serverless-workflow-examples/serverless-workflow-timeouts-showcase/kubernetes/jobs-service-postgresql-knative.yml:
```diff
@@ -0,0 +1,104 @@
+      containers:
+        - name: jobs-service-postgresql
+          image: quay.io/kiegroup/kogito-jobs-service-postgresql-nightly:latest
+          # TODO set to Always?
```
Yes, since it's an online registry with `latest` tag. Set to `IfNotPresent` only if the image is pushed to the internal registry via an external procedure or if your tag has a specific defined version, e.g. `1.24.1`.

wmedvede replied:
And let's go with latest by now, for the future we can see if we want to get the example tight to a particular version.

ricardozanini commented on serverless-workflow-examples/serverless-workflow-timeouts-showcase/kubernetes/jobs-service-postgresql-knative.yml:
```diff
@@ -0,0 +1,104 @@
+    metadata:
+      annotations:
+        autoscaling.knative.dev/minScale: "1"
+        autoscaling.knative.dev/maxScale: "1"
```
I'd instead deploy it as a regular Deployment with `replicas: 1`. I don't think Knative Service is the proper implementation for this use case.

wmedvede replied:
yes, this is something that is open to discussion and proposals. In the technical meeting I tried to introduce this topic as well. With the current job service's implementation, the ideal deployment should be done with a StatefulSet. 
I have spent some time trying to do this setup, but I couldn't make it receive the knative events, it'll try to go back to these experiments, in mean time, we can make progress with this.

To use a regular Deployment is something that we can also do, at the end of the day, behind scene the Knative service approach will make a deployment too, so we'll have the same "singleton" edge cases if we go with a Deployment or a Knative service. However, by using the Knative service we get some benefits for free, like the revisions history, a knative route, etc.

As I said, I'll try to go back to see if its possible to receive knative events when we use the Stateful set. If you have any example of this setup it's very welcome too.

But yeah, this setup is not set in stone.

wmedvede replied:
@ricardozanini the comments above regarding the StatefulSet still apply, but in mean time I've added a jobs-service-postgresql.yml that use a kubernetes deployment.

ricardozanini replied:
@wmedvede what was the problem with the statefulset receiving wiring with knative? The `v1/service` (not knative service), bound to the `statefulset` can't receive incoming knative messages?

wmedvede replied:
Exactly, the statefulstes works a bit different.

When you work with Statefullsets, the pods are not interchangeable, that's a consequence of the fact that the "Pod" represents a stateful application.

Putting it short, imagine that you have a statefulset with two replicas, then, if your statefulset name is stateful-set-example, basically you'll have two pods with fixed names.
    stateful-set-example-0
    stateful-set-example-1
which makes sense, etc., etc.

As a consequence of this, etc., the associated service must be a headless service..., and is not doing exactly any sort of balancing between the pods, since the pods are not interchangeable.

So, even when in our case we have only one pod, when a trigger is configured

Trigger -> stateful-set-example (service) -> the event never arrives to -> stateful-set-example-0 ....

That's the short story :smile: , and the last thing I've tried. However, I don't discard a trick, or any other complementary configuration...
Maybe, considering the nature of the fixed name of the pods, an additional service can be added that refers to the pod fixed name or variants. But paused my investigation in favor of moving all the rest forward.

I'll try to come back to this later, maybe in knative slacks channel they can came up with something.

ricardozanini replied:
@lionelvillard if you have the time to lend your brain :)

lionelvillard replied:
IIRC, using a headless service with selector should work. The load balancing is done at the DNS level so all pods should receive events.

wmedvede replied:
Hello @lionelvillard , many thanks for the pointer.
I don't discard I'm doing, but the example I was trying to configure already has the selector configured. I double-checked, and re-re-tryied and also moved to knative 1.6.0. But my hello world service do not receive the events.


Here's the configuration I was working with in case you can take a look.

https://gist.github.com/wmedvede/039bc2303630daa2f223c048f6831e97

Many thanks again!

ricardozanini commented on serverless-workflow-examples/serverless-workflow-timeouts-showcase/kubernetes/jobs-service-postgresql-knative.yml:
```diff
@@ -0,0 +1,104 @@
+    ref:
+      apiVersion: serving.knative.dev/v1
+      kind: Service
+      name: jobs-service-postgresql
```
If we change to Deployment, we have to update here to look for a `v1/Service` instead.

ricardozanini commented on serverless-workflow-examples/serverless-workflow-timeouts-showcase/timeouts-showcase/src/main/java/org/kie/kogito/examples/EventsProducerResource.java:
```diff
@@ -0,0 +1,162 @@
+                    .withType(eventType)
+                    .withTime(OffsetDateTime.now())
+                    .withExtension("kogitoprocrefid", processInstanceId)
+                    .withData(JsonCloudEventData.wrap(objectMapper.createObjectNode().put("eventData", event.getEventData())))
```
Not sure if this is in the SDK, but it's worth looking at. Knative likes to process the CE with the right `dataContentType` defined. In this case, JSON. We currently have this bug on the runtimes side, which I'll fix in this sprint.

wmedvede replied:
I have executed this code with everything deployed in knative and it worked. 
The scenario was like this:
1) the code above produced the event, and quarkus-http connector wrote it to the K_SINK
2) the event went to the broker
3) the event was "picked-up" by the corresponding trigger
4) the event was delivered to the runtime
5) the process got the event as expected

I'll double check, because I deduce that I should have faced the bug?

wmedvede replied:
btw, I confirm that scenario above works fine with the provided code.

wmedvede replied:
good point,  somewhere in the execution chain, at the cloudevents-json-jackson module the following check is being done     
at serialization/deserialization time, so, if no type, then it assumes json content.

I'll add the content explicitly, will be much reliable.

static boolean dataIsJsonContentType(String contentType) {
        // If content type, spec states that we should assume is json
        return contentType == null || contentType.startsWith("application/json") || contentType.startsWith("text/json");
    }

ricardozanini commented on serverless-workflow-examples/serverless-workflow-timeouts-showcase/pom.xml:
```diff
@@ -0,0 +1,16 @@
+  <version>2.0.0-SNAPSHOT</version>
+  <packaging>pom</packaging>
+
+  <name>Kogito Example :: Serverless Workflow Timeouts Showcase</name>
```
Since we have only one service, why not have it in the parent folder instead of creating a parent pom, then the actual module?

wmedvede replied:
This can be reviewed, I started with this structure because we normally have two modules in this type of examples, kind of one module for the workflow, another module for auxiliary services or the ui, etc. But for now I'll try to keep everything in one module.

wmedvede replied:
Also If I keep the service in the serverless-workflow-timeouts-showcase, the automatically generated names will look a bit longer :( , I'd prefer people to see the knative service name, image name, etc., like "timeouts-showcase" instead of "servlerless-workflow-timeouts-showcase"..., looks a bit long

So to get the shorted name I need to start tunning properties..., which ideally I would like to avoid, I think that the more tunings we add, the more complex to understand it'll be.

ricardozanini replied:
no problem!

ricardozanini reviewed (commented):
Great content; thanks for this PR! I've left a few comments, but my main concern is using KService instead of a Deployment for the Jobs Service. Not sure if `ksvc` is the right implementation for this use case.

wmedvede commented:
@ricardozanini I've added a new file jobs-service-postgresql.yml (  https://github.com/kiegroup/kogito-examples/pull/1374/files#diff-88dae8ad13b384b3b34bf69445dedecd5899cd2fe2597c4a1cd987ba8c73f391 ) which basically uses a kubernetes Deployment and updated the README.md to use this file instead.

Let's to with this...

I'll keep jobs-service-postgresql-knative.yml by now and remove later.

tiagodolphine commented:
@MarianMacik can you take a look and run this example as well, it will be the base for the guide with timeouts + job service. Thanks.

fjtirado commented on serverless-workflow-examples/serverless-workflow-timeouts-showcase/timeouts-showcase/src/main/java/org/kie/kogito/examples/HelperService.java:
```diff
@@ -0,0 +1,33 @@
+/*
```
Do we really need this file?. 
To print traces we should use type custom and operation sysout

tiagodolphine replied:
done

wmedvede replied:
Even when what we are doing is just printing a log, the intention is to emulate a service invocation from the callback state. By doing this, we don't have to create another rest-service, etc.

fjtirado commented on serverless-workflow-examples/serverless-workflow-timeouts-showcase/timeouts-showcase/src/main/resources/callback-state-timeouts.sw.json:
```diff
@@ -0,0 +1,130 @@
+  ],
+  "functions": [
+    {
+      "name": "callbackFunction",
```
To priin traces we should use type custom and operatio sysout

tiagodolphine replied:
done

fjtirado commented on serverless-workflow-examples/serverless-workflow-timeouts-showcase/timeouts-showcase/src/main/resources/callback-state-timeouts.sw.json:
```diff
@@ -0,0 +1,130 @@
+    }
+  ],
+  "errors": [
+    {
```
I think that there is any error, we should allow the flow to fail naturally by not catching it

This flow is equivalent to
`` 
public static void main (String[] args)
try {
}
catch (Exception ex) 
{
 System.out.println(ex.getMessage());
}
}
```
which will return the os a status code of 0, which is not really true

tiagodolphine replied:
done

fjtirado commented on serverless-workflow-examples/serverless-workflow-timeouts-showcase/timeouts-showcase/src/main/resources/callback-state-timeouts.sw.json:
```diff
@@ -0,0 +1,130 @@
+  "functions": [
+    {
+      "name": "callbackFunction",
+      "metadata": {
```
This syntax is outdated
you shouls use 
{
"name": "callbackFunction",
type: "custom",
operation:"service:org.kie.kogito.examples.HelperService:remoteSystemOut"
}

fjtirado replied:
But, again in this case you can perfectly use sysout, there is not need to invoke a java service here

tiagodolphine replied:
removed

fjtirado reviewed (commented):
Workflow needs to be updated to align with other examples when invoking a java service. 
Also I do not think error handling is needed in this example

fjtirado commented:
> @MarianMacik can you take a look and run this example as well, it will be the base for the guide with timeouts + job service. Thanks.

@tiagodolphine The flow shoud be polished a bit

MarianMacik commented on serverless-workflow-examples/serverless-workflow-timeouts-showcase/timeouts-showcase/src/main/resources/switch-state-timeouts.sw.json:
```diff
@@ -0,0 +1,104 @@
+{
+  "id": "switch_state_timeouts",
```
Can we include some tests for this? I think we can make use of the awaitility library to check both workflows and both scenarios (continuation of the workflow by sending an event in less than 30 seconds and waiting for timeout by not sending an event). We could utilize the test utils we have or we no longer want to include them in our examples?

tiagodolphine replied:
yes, I think we can do this in a follow-up PR if we want an integration test to run job-service with knative it would be more difficult unless we just test the runtime with the in-memory job-service.

cristianonicolai replied:
yeah, at least some basic IT test so we can ensure the project runs fine

MarianMacik replied:
@tiagodolphine I think cloud events can be sent also outside of knative/kubernetes, so it should work also locally. We just need to start the jobs service and possibly PostgreSQL (which should be automatically available as a dev service).

tiagodolphine replied:
+1, I'll provide IT for the workflows.

tiagodolphine commented:
@MarianMacik @fjtirado we have 2 options since Walter is on PTO:
1- Merge the PR at it is and I can open a follow-up PR with the changes
2- I can create a new PR rebase with this one and apply the changes
wdyt?
c/c @cristianonicolai

cristianonicolai commented on serverless-workflow-examples/serverless-workflow-timeouts-showcase/timeouts-showcase/pom.xml:
```diff
@@ -0,0 +1,205 @@
+  <parent>
+    <groupId>org.kie.kogito.examples</groupId>
+    <artifactId>serverless-workflow-timeouts-showcase</artifactId>
+    <version>2.0.0-SNAPSHOT</version>
```
```suggestion
    <version>1.0-SNAPSHOT</version>
```

cristianonicolai commented on serverless-workflow-examples/serverless-workflow-timeouts-showcase/timeouts-showcase/pom.xml:
```diff
@@ -0,0 +1,205 @@
+    <project.reporting.outputEncoding>UTF-8</project.reporting.outputEncoding>
+    <quarkus.platform.artifact-id>quarkus-bom</quarkus.platform.artifact-id>
+    <quarkus.platform.group-id>io.quarkus</quarkus.platform.group-id>
+    <quarkus.platform.version>2.11.2.Final</quarkus.platform.version>
```
```suggestion
    <quarkus.platform.version>2.12.0.Final</quarkus.platform.version>
```

cristianonicolai commented:
jenkins please retest this

MarianMacik commented:
@tiagodolphine If you have merging rights you can push to the Walter's branch and it will pop up directly in this PR.

tiagodolphine commented on serverless-workflow-examples/serverless-workflow-timeouts-showcase/timeouts-showcase/src/main/resources/callback-state-timeouts.sw.json:
```diff
@@ -0,0 +1,130 @@
+      "transition": "PrintExitMessage"
+    },
+    {
+      "name": "FinalizeWithError",
```
done

tiagodolphine commented on serverless-workflow-examples/serverless-workflow-timeouts-showcase/kubernetes/jobs-service-postgresql-triggers.yml:
```diff
@@ -0,0 +1,48 @@
+---
```
removed

tiagodolphine commented on serverless-workflow-examples/serverless-workflow-timeouts-showcase/scripts/delete-jobs-service-postgresql.sh:
```diff
@@ -0,0 +1,11 @@
+#!/bin/sh
+
+kubectl delete deployment jobs-service-postgresql
```
this is a helper script (optional) just to make things easier to cleared, we can mention it on the readme

MarianMacik replied:
Yes, that would be great.

tiagodolphine commented on serverless-workflow-examples/serverless-workflow-timeouts-showcase/pom.xml:
```diff
@@ -0,0 +1,16 @@
+  <name>Kogito Example :: Serverless Workflow Timeouts Showcase</name>
+  <modules>
+    <module>timeouts-showcase</module>
+  </modules>
```
I think it was a premature optimization in case there were more services interacting, maybe we can move it to the upper module

cristianonicolai replied:
+1 to move to upper module

tiagodolphine commented on serverless-workflow-examples/serverless-workflow-timeouts-showcase/timeouts-showcase/README.md:
```diff
@@ -0,0 +1,56 @@
+
+If you want to learn more about building native executables, please consult https://quarkus.io/guides/maven-tooling.
+
+## Provided Code
+
+### RESTEasy Reactive
+
+Easily start your Reactive RESTful Web Services
+
+[Related guide section...](https://quarkus.io/guides/getting-started-reactive#reactive-jax-rs-resources)
```
+1

tiagodolphine commented on serverless-workflow-examples/serverless-workflow-timeouts-showcase/timeouts-showcase/src/main/java/org/kie/kogito/examples/GreetingResource.java:
```diff
@@ -0,0 +1,18 @@
+    @GET
+    @Produces(MediaType.TEXT_PLAIN)
+    public String hello() {
+        return "Hello from the Timeouts Showcase: " + ZonedDateTime.now();
```
done

tiagodolphine commented on serverless-workflow-examples/serverless-workflow-timeouts-showcase/README.md:
```diff
@@ -0,0 +1,197 @@
+  -H 'accept: */*' \
+  -H 'Content-Type: application/json' \
+  -d '{
+  "workflowdata": {}
```
+1

tiagodolphine commented on serverless-workflow-examples/serverless-workflow-timeouts-showcase/README.md:
````diff
@@ -0,0 +1,197 @@
+
+```shell
+curl -X 'POST' \
+  'http://timeouts-showcase.default.10.105.86.217.sslip.io/switch_state_timeouts' \
````
I'll add to the readme but it should be a similar behavior just changing the path on the endpoint

tiagodolphine commented:
@MarianMacik @fjtirado applied your comments

cristianonicolai commented on serverless-workflow-examples/serverless-workflow-timeouts-showcase/timeouts-showcase/.gitignore:
```diff
@@ -0,0 +1,40 @@
+#Maven
```
.gitignore, mvnw and mvnw.cmd can be probably removed

tiagodolphine replied:
done

fjtirado reviewed (approved):
Thanks @wmedvede and @tiagodolphine for the quick fix

tiagodolphine commented:
@MarianMacik @cristianonicolai can you take another look? Thanks.

cristianonicolai commented on serverless-workflow-examples/serverless-workflow-timeouts-showcase/src/test/java/org/kie/kogito/examples/TimeoutTest.java:
```diff
@@ -0,0 +1,85 @@
+import static org.awaitility.Awaitility.await;
+import static org.hamcrest.Matchers.equalTo;
+
+@QuarkusTest
```
@tiagodolphine could this be a `@QuarkusIntegrationTest` instead?

tiagodolphine replied:
done

wmedvede commented:
Hi @tiagodolphine , I've just pushed the timeouts showcase UI. Basically a toy, but useful for users to test and play.
Maybe you want to refer/use it in the guide.

![timeouts_showcase](https://user-images.githubusercontent.com/2431454/188923718-820f7d68-2afe-4491-9f23-92465efe4825.png)

ricardozanini commented on serverless-workflow-examples/serverless-workflow-timeouts-showcase/README.md:
````diff
@@ -200,10 +201,65 @@ __  ____  __  _____   ___  __ ____  ______
+failed to call webhook: Post "https://eventing-webhook.knative-eventing.svc:443/sinkbindings?timeout=10s": dial tcp 10.97.131.18:443: connect: connection refused
+```
+
+This is a "minikube + knative quickstart plugin" initialization error, rather than a showcase issue. If this happens in your installation you must remove the knative profile from your minikube installation and repeat the showcase installation procedure.
````
@lionelvillard have you seen this issue before?

wmedvede replied:
@ricardozanini I've opened an issue https://github.com/knative-sandbox/kn-plugin-quickstart/issues/331, and recomendation by now is to reinstall.

radtriste commented on serverless-workflow-examples/serverless-workflow-timeouts-showcase/pom.xml:
```diff
@@ -0,0 +1,202 @@
+    <project.reporting.outputEncoding>UTF-8</project.reporting.outputEncoding>
+    <quarkus.platform.artifact-id>quarkus-bom</quarkus.platform.artifact-id>
+    <quarkus.platform.group-id>io.quarkus</quarkus.platform.group-id>
+    <quarkus.platform.version>2.12.0.Final</quarkus.platform.version>
```
@wmedvede This was updated before your merge: https://github.com/kiegroup/kogito-examples/pull/1390
Can you please update it ?
